ckChangeLog-core/src/main/java/de/cketti/library/changelog/XmlParser.java

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright (C) 2012-2017 cketti and contributors
+ * 
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * 
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package de.cketti.library.changelog;
+
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+import org.xmlpull.v1.XmlPullParser;
+import org.xmlpull.v1.XmlPullParserException;
+
+
+/**
+ * The parser for ckChangeLog's XML file format.
+ */
+public final class XmlParser {
+    private static final int NO_VERSION = -1;
+    private static final String TAG_CHANGELOG = "changelog";
+    private static final String TAG_RELEASE = "release";
+    private static final String ATTRIBUTE_VERSION = "version";
+    private static final String ATTRIBUTE_VERSION_CODE = "versioncode";
+    private static final String TAG_CHANGE = "change";
+
+
+    private final XmlPullParser xmlPullParser;
+    private final int lastVersionCode;
+    private final List<ReleaseItem> result;
+    
+    
+    public static List<ReleaseItem> parse(XmlPullParser xmlPullParser) {
+        XmlParser xmlParser = new XmlParser(xmlPullParser, NO_VERSION);
+        return xmlParser.readChangeLog();
+    }
+
+    public static List<ReleaseItem> parse(XmlPullParser xmlPullParser, int lastVersionCode) {
+        XmlParser xmlParser = new XmlParser(xmlPullParser, lastVersionCode);
+        return xmlParser.readChangeLog();
+    }
+
+    private XmlParser(XmlPullParser xmlPullParser, int lastVersionCode) {
+        this.xmlPullParser = xmlPullParser;
+        this.lastVersionCode = lastVersionCode;
+        result = new ArrayList<>();
+    }
+
+    private List<ReleaseItem> readChangeLog() {
+        try {
+            while (xmlPullParser.getEventType() != XmlPullParser.START_TAG) {
+                xmlPullParser.next();
+            }
+
+            assertElementStart(TAG_CHANGELOG);
+
+            int eventType;
+            do {
+                eventType = xmlPullParser.next(); 
+                if (eventType == XmlPullParser.START_TAG) {
+                    if (parseReleaseElement()) {
+                        // Stop reading more elements if this entry is not newer than the last version.
+                        break;
+                    }
+                }                
+            } while (eventType != XmlPullParser.END_DOCUMENT);
+        } catch (XmlPullParserException | IOException e) {
+            throw new IllegalStateException(e);
+        }
+
+        return result;
```

Comment:
It shouldn't be necessary to create a defensive copy, since the ownership of the list basically goes to the caller (`XmlParser` instances don't outlive the parsing process; see `XmlParser.parse(…)`).

But I guess we could wrap the list using `Collections.unmodifiableList()` to avoid callers depending on a modifiable list.